### ISSUE TITLE
Jitter and MOS calculation

### ISSUE THREAD

**Issue by ishmaelen** [lmangani/rtpsniff#1]
Moving [issue](https://github.com/sipcapture/statstrmr/issues/2) from statstrmr.
- Include Jitter calculation in rtpsniff and use to to calculate a pseudo-MOS.

**Comment by lmangani** [comment]
Pushed some raw test code to `testing` branch needs checks and adjustments on the output formats before implementing the R/MOS estimator... work in progress for testing only

**Comment by ishmaelen** [comment]
Just a quick confirmation it works as expected.
The jitter calculation seems valid and all in all everything works fine.

Added jitter to kamailio.cfg and dashboard.js as well. I've compared it to what e.g. baresip reports; but I'll do more extensive tests tonight after work and compare it with ws/bs. Will post results.

So far so good!

**Comment by lmangani** [comment]
Nice! I'll add the MOS calculation next (I also contributed to the Baresip/Libre [x-rtp](http://lists.creytiv.com/pipermail/re-devel/2014-May/000247.html) stats reporter)

**Comment by lmangani** [comment]
Brutal pseudo-MOS calculation implemented in out_json, mind i did not test this AT ALL yet so most likely far from done.. but its a kickoff ;)

**Comment by ishmaelen** [comment]
Nice!

I will try to look over the MOS stats as well. Basically looks pretty straight forward (the calculation) so shouldn't deviate to much. At least it works right out of the block and prints a value for MOS as is :)

Note the statstrmr parsing adding ' to "name" p-mos.

{ timestamp: 1457460140,
  interval: 10,
  streams: 2,
  packets: 998,
  lost: 2,
  late: 0,
  lost_perc: 0.200401,
  late_perc: 0,
  out_of_seq: 0,
  delay_min: 22490,
  delay_max: 59967,
  delay_avg: 41228,
  jitter: 3742,
  'p-mos': 4,
  type: 'rtp_stat',
  host: 'mrelay01' }
200

**Comment by lmangani** [comment]
Fixed some blind errors on format, please resync

<img src="http://cdn.meme.am/instances2/500x/4684860.jpg" width=200>

**Comment by ishmaelen** [comment]
Jitter results are comparable to baresip/linphone results. I've run some tests on a single node without any other traffic and it looks promising. I believe it's good enough to use as an indication as is.

For MOS have a look at the parts below and pay attention to MOS 5 score in the bottom.

Also, one might question the end relevance for .0000000 perc/mos :)

{ "timestamp": 1457469350, "interval": 10, "streams": 4, "packets": 1994, "lost": 8, "late": 0, "lost_perc": 0.401204, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": 8629, "delay_max": 528351, "delay_avg": 268490, "jitter": 5646, "p-mos": 3.000000, "type": "rtp_stat" }

{ "timestamp": 1457469920, "interval": 10, "streams": 4, "packets": 610, "lost": 2, "late": 0, "lost_perc": 0.327869, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": 35149, "delay_max": 77424, "delay_avg": 56286, "jitter": 5424, "p-mos": 4.000000, "type": "rtp_stat" }

{ "timestamp": 1457469320, "interval": 10, "streams": 6, "packets": 1883, "lost": 8, "late": 0, "lost_perc": **0.424854**, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": **28062**, "delay_max": **2703793**, "delay_avg": **1365927**, "jitter": **6236**, "p-mos": **5.000000**, "type": "rtp_stat" }

**Comment by lmangani** [comment]
Thanks for testing this! the 5 is generated by the range limiter must have been a spike to investigate. pull the latest version to get actual floats for MOS instead of that flat 4 too

**Comment by ishmaelen** [comment]
Yes, it was a spike observed in Firehol/Netdata and [Homer](http://picpaste.com/Homer-FbUR49xo.png). It's easy to generate crap stats.

Now MOS looks good as well! Have a look [here](http://pastebin.com/tnJMEXr9)

{ timestamp: 1457473060,
  interval: 10,
  streams: 4,
  packets: 1076,
  lost: 4,
  late: 0,
  lost_perc: 0.371747,
  late_perc: 0,
  out_of_seq: 0,
  delay_min: 10205,
  delay_max: 120792,
  delay_avg: 65498,
  jitter: 6027,
  'p-mos': 4.07127,             <-- notice how it's 'ed in rtpsniff
  type: 'rtp_stat',
  host: 'mrelay02' }
200

**Comment by lmangani** [comment]
changed p-mos to mos to avoid this, might also have to change float to int*100 to parse in Kamailio
also noticed another issue with delay calculation in case of late packets affecting stats, will check in parallel

**Comment by ishmaelen** [comment]
Checkout these two samples. There is a significant difference in MOS but the numbers are pretty alike except lost_perc; but still insignificantly so. The only variance here is the lost \* 2 part of formula.
Perhaps such a huge factor while there are few lost percentage wise isn't working out optimally?

{ "timestamp": 1457475450, "interval": 10, "streams": 4, "packets": 1992, "lost": 10, "late": 0, "lost_perc": 0.502008, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": 23346, "delay_max": 97782, "delay_avg": 60564, "jitter": 5472, "mos": 3.425179, "type": "rtp_stat" }

{ "timestamp": 1457475460, "interval": 10, "streams": 4, "packets": 1996, "lost": 2, "late": 0, "lost_perc": 0.100200, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": 24006, "delay_max": 95125, "delay_avg": 59565, "jitter": 5677, "mos": 4.242634, "type": "rtp_stat" }

You are correct about Kamailio btw.

**Comment by ishmaelen** [comment]
Sorry for spamming this issue with my ramblings.

I did a quick float to int*100 but that seems to be a "quick and dirty" way of handling float to int?

Might it be better to do something along the line of `intmos = (mos) * (pow(10,6));` (example: 4.386989 \* 10^6 = 4386989) as we know the number of decimals is always 6 and then do something like `int n = (num < 0) ? (num - 0.5) : (num + 0.5);` ?

**flot to int*100**

433 
{ "timestamp": 1457520940, "interval": 10, "streams": 1, "packets": 499, "lost": 1, "late": 0, "lost_perc": 0.200401, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": 16609, "delay_max": 40025, "delay_avg": 28317, "jitter": 2991, "mos": **4.335308**, "type": "rtp_stat" }

426 
{ "timestamp": 1457520960, "interval": 10, "streams": 1, "packets": 498, "lost": 2, "late": 0, "lost_perc": 0.401606, "late_perc": 0.000000, "out_of_seq": 0, "delay_min": 17565, "delay_max": 40023, "delay_avg": 28794, "jitter": 2999, "mos": **4.268798**, "type": "rtp_stat" }

**Comment by lmangani** [comment]
Thanks for the info, keep it coming, brainstorming is king.

I'm not sure we need such granularity on the MOS number, usually its just expressed as %.1f but we can make this an option - for now I'll focus trying to replicate the (loss \* 2.5) discussion from earlier

**Comment by ishmaelen** [comment]
```
    /* Quick and dirty float to int */
    int intmos =(mos * 100);
    // fprintf(stderr, "NEW MOS: %.2f \n", (double)mos);
    mos = intmos;

    if (mos < 0) mos = 0;
    if (mos > 500) mos = 500;
```

git commit out_json.c -m "MOS float to int"
git status
git push origin testing

_remote: Permission to lmangani/rtpsniff.git denied to ishmaelen._

**Comment by lmangani** [comment]
@ishmaelen you need to make a fork and send a pull request to make changes, but I already have the above done will push it tonight

**Comment by lmangani** [comment]
converted double/float to int*100 in out_json plugin

```
{ "timestamp": 1457543930, "interval": 10, "streams": 2, "packets": 500, "lost": 0, "late": 0, "lost_perc": 0, "late_perc": 0, "out_of_seq": 0, "delay_min": 21, "delay_max": 60169, "delay_avg": 30095, "jitter": 2404, "mos": 439, "type": "rtp_stat" }

```

**Comment by lmangani** [comment]
code moved to `master`

**Comment by ishmaelen** [comment]
Just adding a thought - would it be pref. to have a "minimum number of packets" to calculate?

`{ "timestamp": 1458289180, "interval": 10, "streams": 2, "packets": 2, "lost": 0, "late": 0, "lost_perc": 0, "late_perc": 0, "out_of_seq": 0, "delay_min": 0, "delay_max": 0, "delay_avg": 0, "jitter": 0, "mos": 440, "type": "rtp_stat" }`

That isnt very relevant is it..? :)

**Comment by lmangani** [comment]
Its relevant in case sessions starts to die or show up I guess - The purpose of the tool is primarily to detect odd conditions, drops etc so a 0 counts more than a full metric under this perspective IMHO
I suspect those two 1 packet streams were keep-alives or something other than RTP since no ts

**Comment by ishmaelen** [comment]
It was a bad suggestion as it's both easier and better to have such a feature in Homer; e.g. "filter on records with value of field" (ability to filter on <, <=, =, >, >= ...)

PS! It might still be possible to see these reports even with a "Homer filter" if [this feature](https://github.com/sipcapture/homer-ui/issues/66) is implemented.

Sorry for spamming this closed Issue! I'll leave it to history now.
